internal/test/environment/kubernetes/setup.go

```diff
@@ -0,0 +1,352 @@
+package kubernetes
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	volumetypes "github.com/docker/docker/api/types/volume"
+	"github.com/docker/docker/client"
+)
+
+const (
+	// FIXME(vdemeester) those are prefect
+	hyperkubeImage = "gcr.io/google-containers/hyperkube-amd64:v1.8.1"
+	etcdImage      = "quay.io/coreos/etcd:v3.2.9"
+	busyboxImage   = "busybox:latest"
+)
+
+// Setup start a fully containerized kubernetes cluster for e2e tests
+func Setup(ctx context.Context, w io.Writer, kubeconfig string) error {
+	c, err := client.NewEnvClient()
+	if err != nil {
+		return err
+	}
+
+	dinf, err := c.Info(ctx)
+	if err != nil {
+		return err
+	}
+	hostname := dinf.Name
+
+	fmt.Fprintln(w, "Generating certificate authority and kube certificates...")
+	ca, certs, err := generateCertificates(hostname)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintln(w, "Creating kube-config volume...")
+	if err := createKubeConfigVolume(ctx, c, hostname, ca, certs); err != nil {
+		return err
+	}
+
+	fmt.Fprintln(w, "Creating kube-system network...")
+	n, err := c.NetworkCreate(ctx, "kube-system", types.NetworkCreate{})
+	if err != nil {
+		return err
+	}
+
+	starter := containerStarter{c: c, ctx: ctx, network: n.ID, w: w}
+	if err := starter.startContainers([]containerSetup{
+		{
+			name:    "kube-kubelet",
+			cfg:     kubeletContainerConfig(hyperkubeImage, "kubelet.config", ""),
+			hostCfg: kubeletHostConfig,
+		},
+		{
+			name:    "kube-etcd",
+			cfg:     &container.Config{Image: etcdImage, Cmd: []string{"/usr/local/bin/etcd", "-advertise-client-urls=http://0.0.0.0:2379", "-listen-client-urls=http://0.0.0.0:2379"}},
+			hostCfg: &container.HostConfig{RestartPolicy: container.RestartPolicy{Name: "always"}},
+		},
+		{
+			name:    "kube-apiserver",
+			cfg:     apiServerContainerConfig(hyperkubeImage, ""),
+			hostCfg: apiServerHostConfig,
+		},
+		{
+			name:    "kube-controller-manager",
+			cfg:     controllerManagerContainerConfig(hyperkubeImage),
+			hostCfg: controllerManagerHostConfig,
+		},
+		{
+			name:    "kube-scheduler",
+			cfg:     schedulerContainerConfig(hyperkubeImage),
+			hostCfg: schedulerHostConfig,
+		},
+	}); err != nil {
+		return err
+	}
+
+	fmt.Fprintln(w, "Write kubeconfig file...")
+	if err := writeKubeConfig(hostname, kubeconfig, ca); err != nil {
+		return err
+	}
+
+	fmt.Fprintln(w, "Starting kube-dns...")
+	err = withRetries(func() error { return applyYaml(ctx, kubeconfig, kubeDNSYaml) }, 10, 3*time.Second)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintln(w, "Starting weave-net (CNI plugin)...")
+	err = withRetries(func() error { return applyYaml(ctx, kubeconfig, cniYaml) }, 10, 3*time.Second)
```

Comment:
I wonder if we could create a fakeT so that we can use `poll.WaitOn()` for things setup in main.

I think Fail/Fatal could just call exit, and Log would call log.Print()

Reply:
Yeah, I was thinking about that 👼 I'll take a look 😛